### PR TITLE
extracting the observer responsibility to observer class and some ref…

### PR DESCRIPTION
**Pull request by ajitsing**
…actoring

### CODE REVIEW

**Comment by jbarnette** [comment]
@ajitsing Thanks! Can you match the style of the existing code and remove the superficial formatting changes? That'll make it easier to review. I'll mark the spots for cleanup with inline comments.

File: lib/scientist/experiment.rb [hunk 1]
```diff
@@ -190,29 +190,18 @@ def run(name = nil)
     behaviors.freeze
     context.freeze
 
-    name = (name || "control").to_s
+    name = (name || :control).to_s
```
Comment (jbarnette) [inline lib/scientist/experiment.rb]: Is this related to your extraction? If not, please leave it as a String in this diff.

File: lib/scientist/experiment.rb [hunk 2]
```diff
@@ -190,29 +190,18 @@ def run(name = nil)
...
-    end
-
-    unless should_experiment_run?
-      return block.call
-    end
-
-    if @_scientist_before_run
-      @_scientist_before_run.call
-    end
+    missing_behaviour(name) if block.nil?
```
Comment (jbarnette) [inline lib/scientist/experiment.rb]: It's spelled "behavior" everywhere else in the code, including this file.

File: lib/scientist/experiment.rb [hunk 3]
```diff
@@ -190,29 +190,18 @@ def run(name = nil)
...
 
-    observations = []
+    return block.call unless should_experiment_run?
 
-    behaviors.keys.shuffle.each do |key|
-      block = behaviors[key]
-      observations << Scientist::Observation.new(key, self, &block)
-    end
+    @_scientist_before_run.call if @_scientist_before_run
```
Comment (jbarnette) [inline lib/scientist/experiment.rb]: Please change this back to an `if` block, not a suffix.

File: lib/scientist/experiment.rb [hunk 4]
```diff
@@ -226,11 +215,7 @@ def run(name = nil)
       raise MismatchError.new(self.name, result)
     end
 
-    if control.raised?
-      raise control.exception
-    else
-      control.value
-    end
+    control.raised? ? raise(control.exception) : control.value
```
Comment (jbarnette) [inline lib/scientist/experiment.rb]: Please change this back to an `if` / `else` block, not a ternary.

File: lib/scientist/experiment.rb [hunk 5]
```diff
@@ -282,4 +267,9 @@ def raise_on_mismatches?
       !!raise_on_mismatches
     end
   end
+
+  private
+  def missing_behaviour(behaviour)
```
Comment (jbarnette) [inline lib/scientist/experiment.rb]: Please inline this `raise` instead of extracting a `missing_behavior` method.

File: test/scientist/observer_test.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
+describe Scientist::Observer do
+
+  before do
+    block1 = lambda {:block1}
+    block2 = lambda {:block2}
+    @behaviours = {:behaviour1 => block1, :behaviour2 => block2}
+    @experiment = Scientist::Experiment.new 'test'
+  end
+
+  it 'observes the behaviours and records observations' do
```
Comment (jbarnette) [inline test/scientist/observer_test.rb]: Please use `"` consistently.

File: lib/scientist/observer.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+      @observations << Scientist::Observation.new(key, @experiment, &block)
+    end
+    @observations
+  end
+
+  def observation(name)
+    @observations ||= observe
+    @observations.detect { |o| o.name == name }
+  end
+end
```
Comment (jbarnette) [inline lib/scientist/observer.rb]: Please end all files with a newline.

**Comment by ajitsing** [comment]
Thanks for the quick feedback. It has been taken care of.

**Comment by jbarnette** [comment]
@ajitsing Thank you! I'll review this behavior change sometime in the next few days.

**Comment by ajitsing** [comment]
@jbarnette any update on this?

**Comment by ajitsing** [comment]
Removed "minor refactoring => (if context.nil? -> unless context.nil?)" commit from PR.

**Comment by rick** [comment]
Was there a force push here or something?

**Comment by ajitsing** [comment]
@rick Yes, I pushed a commit to this pull request by mistake.. So to remove that commit I had to delete the commit from my branch.

**Comment by rick** [comment]
This seems to have languished (sorry, was probably on our maintaining end) -- bumping for great justice.

**Comment by jbarnette** [comment]
I'm doing some long-deferred PR and issue maintenance, but with fresh eyes I'm not seeing a lot of benefit from this change. 🙇 @ajitsing I'm going to close this, but I'm happy to reopen and look at applying if you have a specific problem in mind that this solves.
